### PR TITLE
NIFI-8025 - Refactored AccessResource SAML and OIDC endpoints into separate classes

### PR DESCRIPTION
**Pull request by thenatog**
NIFI-8025 - Moved SAML to separate class and tested to be working.

NIFI-8025 - Added license to new files.

NIFI-8025 - Fixed a SAML logout consumer path.

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/AccessResource.java [hunk 1]
```diff
@@ -170,16 +105,35 @@
...
+//    @Path("oidc/logoutCallback")
+//    @ApiOperation(
+//            value = "Redirect/callback URI for processing the result of the OpenId Connect logout sequence.",
+//            notes = NON_GUARANTEED_ENDPOINT
+//    )
+//    public void oidcLogoutCallback(@Context HttpServletRequest httpServletRequest, @Context HttpServletResponse httpServletResponse) throws Exception {
+//        // only consider user specific access over https
+//
+//    }
+
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/AccessResource.java:136]: Is there a reason for adding this commented section?  It looks like these lines could be removed.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+        RequestConfig config = RequestConfig.custom()
+                .setConnectTimeout(msTimeout)
+                .setConnectionRequestTimeout(msTimeout)
+                .setSocketTimeout(msTimeout)
+                .build();
+
+        CloseableHttpClient httpClient = HttpClientBuilder
+                .create()
+                .setDefaultRequestConfig(config)
+                .build();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java:455]: What do you think about refactoring HttpClient to a class variable instead of recreating it on every method invocation?  It might also be worth considering replacing it with OkHttp, but that could be handled in a separate task.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/SAMLAccessResource.java [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+        if (!httpServletRequest.isSecure()) {
+            forwardToLoginMessagePage(httpServletRequest, httpServletResponse, AUTHENTICATION_NOT_ENABLED_MSG);
+            return;
+        }
+
+        // ensure saml is enabled
+        if (!samlService.isSamlEnabled()) {
+            forwardToLoginMessagePage(httpServletRequest, httpServletResponse, SAMLService.SAML_SUPPORT_IS_NOT_CONFIGURED);
+            return;
+        }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/SAMLAccessResource.java:209]: This logic is repeated in several other methods, what do you think about refactoring it to a shared method that returns a boolean to determine whether to proceed?  Perhaps something along the lines of `boolean isSamlEnabled(HttpServletRequest, HttpServletResponse)`?

**Review by exceptionfactory — COMMENTED** [review]
Thanks for refactoring the SAML and OIDC resources @thenatog, this should make it easier to maintain these features down the road.  There may be some additional opportunities for improvement, and I realize most of this is just reorganizing things.  There may be an opportunity to reduce some code duplication when checking for request security and method enabled status, but otherwise this seems straightforward.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/saml/impl/StandardSAMLStateManager.java [hunk 4]
```diff
@@ -140,4 +140,7 @@ public String getJwt(final String requestIdentifier) {
         }
     }
 
+    public void setJwtService(JwtService jwtService) {
+        this.jwtService = jwtService;
+    }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/saml/impl/StandardSAMLStateManager.java:145]: Is this change necessary?  It looks like the JwtService is provided through a constructor argument as part of the existing Spring configuration.

**Comment by thenatog** [comment]
Updated PR to address latest feedback.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+
+        HttpPost httpPost = new HttpPost(revokeEndpoint);
+
+        List<NameValuePair> params = new ArrayList<>();
+        // Append a query param with the access token
+        params.add(new BasicNameValuePair("token", accessToken));
+        httpPost.setEntity(new UrlEncodedFormEntity(params));
+
+        try (CloseableHttpResponse response = httpClient.execute(httpPost)) {
+            httpClient.close();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java:464]: This line will close the `httpClient`, but now that it is a class member variable, it should not be closed in this method.  Creating a new method annotated with `@PreDestroy` is one option for closing the HttpClient instance prior to application shutdown.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java (hunk 5 above)
Comment (thenatog) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/OIDCAccessResource.java:464]: Tried to commit this change but somehow 'Merge branch 'main' into NIFI-8025-rebased' happened. Will fix.

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the changes @thenatog! Verified SAML and OIDC with Keycloak, looks good! +1 Merging.
